File: src/stores/settings-store.ts
import { create } from 'zustand';
import AsyncStorage from '@react-native-async-storage/async-storage';

const SETTINGS_KEY = 'plantapp:settings';

export type TemperatureUnit = 'celsius' | 'fahrenheit';
export type LengthUnit = 'cm' | 'in';

interface SettingsState {
  temperatureUnit: TemperatureUnit;
  lengthUnit: LengthUnit;
  notificationsEnabled: boolean;
  isLoaded: boolean;

  load: () => Promise<void>;
  setTemperatureUnit: (unit: TemperatureUnit) => void;
  setLengthUnit: (unit: LengthUnit) => void;
  setNotificationsEnabled: (enabled: boolean) => void;
}

interface PersistedSettings {
  temperatureUnit: TemperatureUnit;
  lengthUnit: LengthUnit;
  notificationsEnabled: boolean;
}

async function persist(settings: PersistedSettings): Promise<void> {
  await AsyncStorage.setItem(SETTINGS_KEY, JSON.stringify(settings));
}

export const useSettingsStore = create<SettingsState>((set, get) => ({
  temperatureUnit: 'celsius',
  lengthUnit: 'cm',
  notificationsEnabled: true,
  isLoaded: false,

  load: async () => {
    try {
      const raw = await AsyncStorage.getItem(SETTINGS_KEY);
      if (raw) {
        const parsed = JSON.parse(raw) as Partial<PersistedSettings>;
        set({
          temperatureUnit: parsed.temperatureUnit ?? 'celsius',
          lengthUnit: parsed.lengthUnit ?? 'cm',
          notificationsEnabled: parsed.notificationsEnabled ?? true,
          isLoaded: true,
        });
      } else {
        set({ isLoaded: true });
      }
    } catch {
      set({ isLoaded: true });
    }
  },

  setTemperatureUnit: (unit) => {
    set({ temperatureUnit: unit });
    const s = get();
    persist({ temperatureUnit: unit, lengthUnit: s.lengthUnit, notificationsEnabled: s.notificationsEnabled });
  },

  setLengthUnit: (unit) => {
    set({ lengthUnit: unit });
    const s = get();
    persist({ temperatureUnit: s.temperatureUnit, lengthUnit: unit, notificationsEnabled: s.notificationsEnabled });
  },

  setNotificationsEnabled: (enabled) => {
    set({ notificationsEnabled: enabled });
    const s = get();
    persist({ temperatureUnit: s.temperatureUnit, lengthUnit: s.lengthUnit, notificationsEnabled: enabled });
  },
}));

File: lib/stores/settings_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

// Settings state
class SettingsState {
  final String temperatureUnit; // 'celsius' | 'fahrenheit'
  final String lengthUnit; // 'cm' | 'in'
  final bool notificationsEnabled;
  final bool isLoaded;

  const SettingsState({
    this.temperatureUnit = 'celsius',
    this.lengthUnit = 'cm',
    this.notificationsEnabled = true,
    this.isLoaded = false,
  });

  SettingsState copyWith({
    String? temperatureUnit,
    String? lengthUnit,
    bool? notificationsEnabled,
    bool? isLoaded,
  }) =>
      SettingsState(
        temperatureUnit: temperatureUnit ?? this.temperatureUnit,
        lengthUnit: lengthUnit ?? this.lengthUnit,
        notificationsEnabled: notificationsEnabled ?? this.notificationsEnabled,
        isLoaded: isLoaded ?? this.isLoaded,
      );
}

class SettingsNotifier extends StateNotifier<SettingsState> {
  SettingsNotifier() : super(const SettingsState());

  static const _tempKey = 'temperatureUnit';
  static const _lengthKey = 'lengthUnit';
  static const _notifKey = 'notificationsEnabled';

  Future<void> load() async {
    final prefs = await SharedPreferences.getInstance();
    state = SettingsState(
      temperatureUnit: prefs.getString(_tempKey) ?? 'celsius',
      lengthUnit: prefs.getString(_lengthKey) ?? 'cm',
      notificationsEnabled: prefs.getBool(_notifKey) ?? true,
      isLoaded: true,
    );
  }

  Future<void> setTemperatureUnit(String unit) async {
    state = state.copyWith(temperatureUnit: unit);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tempKey, unit);
  }

  Future<void> setLengthUnit(String unit) async {
    state = state.copyWith(lengthUnit: unit);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_lengthKey, unit);
  }

  Future<void> setNotificationsEnabled(bool enabled) async {
    state = state.copyWith(notificationsEnabled: enabled);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_notifKey, enabled);
  }
}

final settingsProvider =
    StateNotifierProvider<SettingsNotifier, SettingsState>((ref) {
  return SettingsNotifier();
});
